ext/app/pipe/components.js

```diff
@@ -0,0 +1,322 @@
+(function() {
+
+  // ##################### CSV VIEWER #####################
+
+  // A custom webcomponent that can be used to embed a CSV or Grist file in a page.
+  // For example <csv-viewer href="csv or grist file url"></csv-viewer>
+  // Though the name is csv-viewer, it can also be used to view grist files, accepts all the parameters
+  // from the bootstrapGrist function and a href attribute that will be used as initial-file if it ends with .grist
+  // or initial-data if it ends with .csv
+  style(`
+    csv-viewer {
+      display: block;
+    }
+  `);
+  // Now build the custom element.
+  class CsvViewer extends HTMLElement {
+    constructor() {
+      super();
+
+    }
+    connectedCallback() {
+      if (this.initiated) {return;}
+      this.initiated = true;
+      // Also support the href attribute, which is easier to use.
+      const href = this.getAttribute('href');
+      const csvHref = href?.endsWith('.csv') ? href : undefined;
+      const gristHref = href?.endsWith('.grist') ? href : undefined;
+      const initialFile = this.getAttribute('initial-file') || gristHref;
+      const name = this.getAttribute('name') || "";
+      const initialData = this.getAttribute('initial-data') || csvHref;
+      const initialContent = this.getAttribute('initial-content');
+      const singlePage = this.hasAttribute('single-page');
+      const loader = this.hasAttribute('loader');
+      bootstrapGrist({
+        element: this,
+        name,
+        initialFile,
+        initialData,
+        initialContent,
+        singlePage,
+        loader
+      });
+    }
+  }
+  customElements.define('csv-viewer', CsvViewer);
+
+
+  // ##################### Open with Grist popup #####################
+
+  // Adds a global click handler for elements that should open a Grist document in a popup.
+
+  window.addEventListener('click', (clickEvent) => {
+    if (!clickEvent.target.matches('[data-grist-csv-open],[data-grist-doc-open]')) {
+      return;
+    }
+    clickEvent.preventDefault();
+    previewInGristClickHandler(clickEvent.target);
+  });
+
+
+  function previewInGrist(options) {
```

Comment:
We have some documentation for `bootstrapGrist` in the `README.md`, and also in the website maintained in the `docs/` subdirectory.

Could you add a new section to the README for the CSV viewer, that covers `previewInGrist`, attribute handling, and the custom element?